rust/monovertex/src/transformer.rs

```diff
@@ -1,61 +1,160 @@
-use crate::error::Result;
-use crate::message::Message;
+use std::collections::HashMap;
+
+use crate::config::config;
+use crate::error::{Error, Result};
+use crate::message::{Message, Offset};
 use crate::shared::utc_from_timestamp;
-use crate::sourcetransform_pb::source_transform_client::SourceTransformClient;
-use crate::sourcetransform_pb::SourceTransformRequest;
+use crate::sourcetransform_pb::{
+    self, source_transform_client::SourceTransformClient, SourceTransformRequest,
+    SourceTransformResponse,
+};
+
+use tokio::sync::mpsc;
+use tokio::task::JoinHandle;
+use tokio::time::Duration;
+use tokio_stream::wrappers::ReceiverStream;
+use tokio_util::sync::CancellationToken;
 use tonic::transport::Channel;
+use tonic::{Request, Streaming};
 
 const DROP: &str = "U+005C__DROP__";
 
 /// TransformerClient is a client to interact with the transformer server.
-#[derive(Clone)]
+#[derive(Debug)]
 pub struct SourceTransformer {
-    client: SourceTransformClient<Channel>,
+    read_tx: mpsc::Sender<SourceTransformRequest>,
+    resp_stream: Streaming<SourceTransformResponse>,
 }
 
 impl SourceTransformer {
-    pub(crate) async fn new(client: SourceTransformClient<Channel>) -> Result<Self> {
-        Ok(Self { client })
+    pub(crate) async fn new(mut client: SourceTransformClient<Channel>) -> Result<Self> {
+        let (read_tx, read_rx) = mpsc::channel(config().batch_size as usize);
+        let read_stream = ReceiverStream::new(read_rx);
+
+        // do a handshake for read with the server before we start sending read requests
+        let handshake_request = SourceTransformRequest {
+            request: None,
+            handshake: Some(sourcetransform_pb::Handshake { sot: true }),
+        };
+        read_tx.send(handshake_request).await.map_err(|e| {
+            Error::TransformerError(format!("failed to send handshake request: {}", e))
+        })?;
+
+        tracing::info!("Sending stream request");
+        let transform_fn_with_timeout = tokio::time::timeout(
+            Duration::from_secs(3),
+            client.source_transform_fn(Request::new(read_stream)),
+        );
+        let Ok(resp_stream) = transform_fn_with_timeout.await else {
+            return Err(Error::TransformerError(
+                "connection to transformer gRPC server timed out".to_string(),
+            ));
+        };
```

Comment:
only 3 sec timeout? will this whole thing become flaky in a busy system?

Reply:
Why do we need timeout?

Reply:
Since we create the stream at the beginning and we don't create multiple streams, this should succeed immediately. I added the timeout because the tests were stuck when there was some issue with the server (server started,but there was a mismatch  in the protobuf files on the client and the server). I will remove the timeouts.